runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
+     * also be used in terminal measurements where the measurement processes below are unsupported.
      *
-     * @note This is backed by the `Catalyst::Runtime::CacheManager<ComplexT>` property in
-     * the device implementation.
+     * @param wire The qubit to measure.
+     * @param postselect Optional parameter to force the result to the provided state (roughly
+     *                   equivalent to post-selection).
+     *
+     * @return `Result` The measurement result.
      */
-    virtual void StartTapeRecording() = 0;
+    virtual auto Measure(QubitIdType wire, std::optional<int32_t> postselect) -> Result = 0;
 
     /**
-     * @brief Stop recording a quantum tape if provided.
+     * @brief (Optional) Apply an arbitrary Unitary matrix to the device.
      *
-     * @note This is backed by the `Catalyst::Runtime::CacheManager<ComplexT>` property in
-     * the device implementation.
+     * Instead of identifying an operation by name, this instruction uses the mathematical
+     * representation of a quantum operator as a Unitary matrix in the computational basis.
+     *
+     * See `NamedOperation` for additional gate semantics.
+     *
+     * @param matrix A 1D array representation of the matrix in row-major format.
+     * @param wires Qubits to apply the operation to.
+     * @param inverse Apply the inverse of the operation.
+     * @param controlled_wires Control qubits applied to the operation.
+     * @param controlled_values Control values associated to the control qubits (equal length).
      */
-    virtual void StopTapeRecording() = 0;
+    virtual void
+    MatrixOperation([[maybe_unused]] const std::vector<std::complex<double>> &matrix,
+                    [[maybe_unused]] const std::vector<QubitIdType> &wires, bool inverse = false,
+                    [[maybe_unused]] const std::vector<QubitIdType> &controlled_wires = {},
+                    [[maybe_unused]] const std::vector<bool> &controlled_values = {})
+    {
+        RT_FAIL("MatrixOperation is unsupported by device");
+    }
 
     /**
-     * @brief Result value for "Zero" used in the measurement process.
+     * @brief (Optional) Initialize qubits to a computational basis state.
      *
-     * @return `Result`
+     * This instruction initializes a set of qubits to the provided computational basis state.
+     * Although the instruction is typically used for state initialization at the beginning of
+     * a quantum program, devices are encouraged to support arbitrary re-initialization if
+     * capable. Reinitialization on a subset of qubits is equivalent to deallocation -> allocation
+     * -> initialization on the same set of qubits.
+     * See `ReleaseQubit` for caveats around releasing / resetting entangled qubits.
+     *
+     * @param n Bitstring representation of the basis state |n>, stored as a Byte-array.
+     * @param wires The qubits to initialize.
      */
```

Comment:
```suggestion
     *
     * Devices that lack a `setBasisState` implementation should take care of decomposing `qml.BasisState`.
     */
```

Feel free to ignore this, not sure how I feel about having this piece of frontend note in the runtime.